l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,478 @@
+    event NodeDeactivated(address indexed nodeOwner);
+    event NodeActivated(address indexed nodeOwner);
+    event NodeRemoved(address indexed nodeOwner);
+    event NodeValidatorWeightChanged(address indexed nodeOwner, uint32 newWeight);
+    event NodeAttesterWeightChanged(address indexed nodeOwner, uint32 newWeight);
+    event NodeValidatorPubKeyChanged(address indexed nodeOwner, BLS12_381PublicKey newPubKey);
+    event NodeValidatorPoPChanged(address indexed nodeOwner, BLS12_381Signature newPoP);
+    event NodeAttesterPubKeyChanged(address indexed nodeOwner, Secp256k1PublicKey newPubKey);
+    event ValidatorsCommitted();
+    event AttestersCommitted();
+
+    modifier onlyOwnerOrNodeOwner(address _nodeOwner) {
+        if (owner() != msg.sender && _nodeOwner != msg.sender) {
+            revert UnauthorizedOnlyOwnerOrNodeOwner();
+        }
+        _;
+    }
+
+    constructor(address _initialOwner) {
+        if (_initialOwner == address(0)) {
+            revert InvalidInputNodeOwnerAddress();
+        }
+        _transferOwnership(_initialOwner);
+    }
+
+    /// @notice Adds a new node to the registry.
+    /// @dev Fails if node owner already exists.
+    /// @dev Fails if a validator with the same public key already exists.
+    /// @dev Fails if an attester with the same public key already exists.
+    /// @param _nodeOwner The address of the new node's owner.
+    /// @param _validatorWeight The voting weight of the validator.
+    /// @param _validatorPubKey The BLS12-381 public key of the validator.
+    /// @param _validatorPoP The proof-of-possession (PoP) of the validator's public key.
+    /// @param _attesterWeight The voting weight of the attester.
+    /// @param _attesterPubKey The ECDSA public key of the attester.
+    function add(
+        address _nodeOwner,
+        uint32 _validatorWeight,
+        BLS12_381PublicKey calldata _validatorPubKey,
+        BLS12_381Signature calldata _validatorPoP,
+        uint32 _attesterWeight,
+        Secp256k1PublicKey calldata _attesterPubKey
+    ) external onlyOwner {
+        verifyInputAddress(_nodeOwner);
+        verifyInputBLS12_381PublicKey(_validatorPubKey);
+        verifyInputBLS12_381Signature(_validatorPoP);
+        verifyInputSecp256k1PublicKey(_attesterPubKey);
+
+        uint256 len = nodeOwners.length;
+        for (uint256 i = 0; i < len; ++i) {
+            if (nodeOwners[i] == _nodeOwner) {
+                revert NodeOwnerAlreadyExists();
+            }
+            if (compareBLS12_381PublicKey(nodes[nodeOwners[i]].validatorLatest.pubKey, _validatorPubKey)) {
+                revert ValidatorPubKeyAlreadyExists();
+            }
+            if (compareSecp256k1PublicKey(nodes[nodeOwners[i]].attesterLatest.pubKey, _attesterPubKey)) {
+                revert AttesterPubKeyAlreadyExists();
+            }
```

Comment:
Please remove `O(n)` complexity from this function. 
1) Instead of iterating over `nodes` array just check whether `nodes[_nodeOwner]` is empty.
2) Instead of checking `validatorLatest.pubKey` and `attesterLatest.pubKey` uniquness iterating over `nodes` array just create two mapping in storage `validatorsPubKeyHashes` and `attestePubKeyHashes` and set the mapping value with `true` when pubkey is added as active

Reply:
Changes here: https://github.com/matter-labs/era-contracts/pull/555/commits/e2ea3de6a8c65fa142012bb5360add38a9288431